M2/Macaulay2/m2/computations.m2

```diff
@@ -0,0 +1,139 @@
+fetchComputation = method(Options => true)
+fetchComputation(Type, HashTable,            Context) := Computation => true >> opts -> (C, T,    context) -> fetchComputation(C, T, T, context)
+fetchComputation(Type, HashTable, Sequence,  Context) :=
+fetchComputation(Type, HashTable, HashTable, Context) := Computation => true >> opts -> (C, T, X, context) -> (
+    -- TODO: look for other compatible contexts as well
+    -- TODO: use https://github.com/Macaulay2/M2/issues/1596 when it is implemented
+    if T.cache#?context then T.cache#context
+    else T.cache#context = new C from X)
+
+-----------------------------------------------------------------------------
+-- isComputationDone
+-----------------------------------------------------------------------------
+
+-- determines whether further computation is necessary
+isComputationDone = method(TypicalValue => Boolean, Options => true)
+-- the default method only checks whether Result is non-null
+isComputationDone Computation := Boolean => true >> opts -> container -> container.Result =!= null
+-- TODO: use isReady?
+
+-----------------------------------------------------------------------------
+-- updateComputation
+-----------------------------------------------------------------------------
+
+-- update the computation with the given result
+updateComputation = method(Options => true)
+updateComputation(Computation, Thing)   := true >> opts -> (container, result) -> container.Result = result
+updateComputation(Computation, Nothing) := true >> opts -> (container, result) -> null
+
+-----------------------------------------------------------------------------
+-- adjustComputation
+-----------------------------------------------------------------------------
+
+-- adjust and return the cached result as appropriate based on opts
+adjustComputation = method(Options => true)
+adjustComputation Computation := true >> opts -> container -> container.Result
+
+-----------------------------------------------------------------------------
+-- cacheComputation
+-----------------------------------------------------------------------------
+
+-- perform the computation, or return the cached result
+cacheComputation = method(TypicalValue => CacheFunction, Options => true)
+-- the default method only sets the Result in the container
+-- Note: this function takes advantage of FunctionClosures by modifying the container
+cacheComputation Computation := CacheFunction => true >> opts -> container -> new CacheFunction from (
+    algorithm -> (
+	if isComputationDone(opts, container) then ( cacheHit container;
+	    adjustComputation(opts, container))
+	else updateComputation(opts, container, algorithm(opts, container))))
+
+-----------------------------------------------------------------------------
+-- Introspective tools for Computations
+-----------------------------------------------------------------------------
+
+-- toggle printing debug information or collecting data about cache hits
```

Comment:
This section is not fully developed because I wasn't sure what methods of introspection would be useful. I sort of thought about having a computation specific debug level, so it would take values like 0 1 2 3. I'm happy to comment it out for now and add it when it's more polished.

Reply:
Also if you have suggestions about good ways to look into the state of a computation, I'd love to hear them!

Reply:
I updated this code to xor the debug level with 1, changing it's parity. I'll explain below.